### PR TITLE
feat: implement `Logical` intrinsic

### PR DESCRIPTION
**Pull request by shivimathur**
Fixes : https://github.com/lfortran/lfortran/issues/3690

### CODE REVIEW

**Comment by shivimathur** [comment]
``` fortran
logical(kind = 8)::y = .true.
```
This does not change kind of y from 4 to 8.

File: a.f90 [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    print*, kind(logical(.true., 8))
+    print*, logical(x, 4)
+    if (logical(x, 4) .neqv. .true.) error stop
+
+    print*, logical(y)
+    if (logical(y) .neqv. .false.) error stop
+
+    print*, kind(logical(x, 8))
+    print*, kind(logical(y))
+end
```
Comment (Pranavchiku) [inline a.f90:18]: you can remove this file and register this as an integration test

File: src/libasr/pass/intrinsic_functions.h [hunk 2]
```diff
@@ -1743,6 +1744,30 @@ namespace Ibits {
...
+    static ASR::expr_t *eval_Logical(Allocator &al, const Location &loc,
+            ASR::ttype_t* t1, Vec<ASR::expr_t*> &args, diag::Diagnostics& /*diag*/) {
+        bool result = ASR::down_cast<ASR::LogicalConstant_t>(args[0])->m_value;
+        return make_ConstantWithType(make_LogicalConstant_t, result, t1, loc);
+    }
+
+    static inline ASR::expr_t* instantiate_Logical(Allocator &al, const Location &loc,
+            SymbolTable *scope, Vec<ASR::ttype_t*>& arg_types, ASR::ttype_t *return_type,
+            Vec<ASR::call_arg_t>& new_args, int64_t /*overload_id*/) {
+        declare_basic_variables("_lcompilers_optimization_logical_" + type_to_str_python(arg_types[0]));
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_functions.h:1758]:
```suggestion
        declare_basic_variables("_lcompilers_logical_" + type_to_str_python(arg_types[0]));
```

File: src/libasr/pass/intrinsic_functions.h [hunk 3]
```diff
@@ -1743,6 +1744,30 @@ namespace Ibits {
...
+        fill_func_arg("x", arg_types[0]);
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_functions.h:1759]:
```suggestion
        fill_func_arg("l", arg_types[0]);
```

**Review by Pranavchiku — CHANGES_REQUESTED** [review]
Left a few comments, rest all is good.

**Comment by Shaikh-Ubaid** [comment]
>I don't think we should be adding intrinsics for simple casts like this one. Cannot we simply insert a regular cast node here?

I think we should simply insert a cast node.

**Comment by HarshitaKalani** [comment]
>I don't think we should be adding intrinsics for simple casts like this one. Cannot we simply insert a regular cast node here?

I agree to this.
@shivimathur , can you try implementing `logical` as a cast node?
For reference, you can follow: https://github.com/lfortran/lfortran/pull/3443/files

**Comment by Pranavchiku** [comment]
> I don't think we should be adding intrinsics for simple casts like this one. Cannot we simply insert a regular cast node here?

I agree on this, I commented at https://github.com/lfortran/lfortran/pull/3922#issuecomment-2078688167

**Pull request closed without merging**